### PR TITLE
Mkldnn with generic fallback backend

### PR DESCRIPTION
**Pull request by okdshin**
This PR adds new backend that  uses MKLDNN and "Generic" (i.e. pure C++ implementation). When op is not supported by MKLDNN, Generic is used (fallback).
FYI: This PR contains fallback system only so operators must be added.
This PR depends on PR #94.

### CODE REVIEW

**Comment by okdshin** [comment]
I would like to add generic test which enable to test across backends.
I will add it by another PR.

File: menoh/mkldnn_with_generic_fallback/backend/mkldnn/mkldnn_context.cpp [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+                    variable_memory_table_.insert(
+                      std::make_move_iterator(new_output_memory_list.begin()),
+                      std::make_move_iterator(new_output_memory_list.end()));
+                    temp_memory_list_.insert(
+                      temp_memory_list_.end(),
+                      std::make_move_iterator(new_temp_memory_list.begin()),
+                      std::make_move_iterator(new_temp_memory_list.end()));
+                }
+
+                // when any nodes are not processed
```
Comment (msakai) [inline menoh/mkldnn_with_generic_fallback/backend/mkldnn/mkldnn_context.cpp]: "when no nodes were processed" might be better.

Comment (okdshin) [reply menoh/mkldnn_with_generic_fallback/backend/mkldnn/mkldnn_context.cpp]: I'll fix. Thanks.

File: menoh/mkldnn_with_generic_fallback/backend/mkldnn/mkldnn_context.cpp [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+                                    std::tie(copy_proc, arr) = *found;
+                                    assert(arr.dims().size() == 2 ||
+                                           arr.dims().size() == 4);
+                                    new_copy_procedure_list.push_back(
+                                      copy_proc);
+                                    input_list.push_back(arr);
+                                    auto format =
+                                      arr.dims().size() == 2
+                                        ? mkldnn::memory::format::nc
+                                        : mkldnn::memory::format::nchw;
```
Comment (msakai) [inline menoh/mkldnn_with_generic_fallback/backend/mkldnn/mkldnn_context.cpp]: It seems that the `format` variable is unused.

File: menoh/mkldnn_with_generic_fallback/context.hpp [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+namespace menoh_impl {
+    namespace mkldnn_with_generic_fallback_backend {
+
+        class context {
+        public:
+            optional<std::tuple<procedure, array>>
+            try_to_get_variable(std::string const& name) {
+                return do_try_to_get_variable(name);
+            }
```
Comment (msakai) [inline menoh/mkldnn_with_generic_fallback/context.hpp]: `try_to_get_variable` returns a procedure for copying, but I couldn't understand from where to where.
Backend-specific memory (e.g. GPU memory) to normal CPU memory?
Also, are those backend-specific memory managed in the array?, or managed separately inside backends?

Comment (okdshin) [reply menoh/mkldnn_with_generic_fallback/context.hpp]: `try_to_get_variable` returns a procedure for copying from Backend-specific memory to normal CPU memory.
Those backend-specific memories are managed separately inside backends.

File: menoh/mkldnn_with_generic_fallback/backend/mkldnn/mkldnn_context.cpp (hunk 2 above)
Comment (okdshin) [reply menoh/mkldnn_with_generic_fallback/backend/mkldnn/mkldnn_context.cpp]: I see. I'll fix.

**Comment by msakai** [comment]
I think current API is enough for now and this is just a remark:

`menoh_impl::mkldnn_with_generic_fallback_backend::process_node_list` processes the node list linearly and returns when it first encounters _unsupported_ operator. It is a simple API but might not be optimal, because there might be a supported node after the unsupported node and the node does not depend on outputs of the unsupported node.

**Comment by okdshin** [comment]
I see. It will be a problem when we try to load complex models which have branches in execution path. Thanks for your pointing out.

**Comment by okdshin** [comment]
This PR lacks the test to send variables from MKLDNN backend to Generic backend.
#106 enables to make such tests easily so we should merge #106 first.

**Comment by okdshin** [comment]
#106 failed ApVeyor CI so I merge this previously.

**Pull request merged**